kernel: userspace: aligned memory allocation for dynamic objects

dcpleung opened the pull request:
This allows allocating dynamic kernel objects with memory alignment
requirements. The first candidate is for thread objects for x86
and x86_64 where saving/restoring SSE/FPU registers require strict
alignment.
    
Fixes #29589
Fixes #29629

dcpleung committed:
kernel: mempool: add z_thread_aligned_alloc

This adds a new z_thread_aligned_alloc() to do memory allocation
with required alignment.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: userspace: aligned memory allocation for dynamic objects

This allows allocating dynamic kernel objects with memory alignment
requirements. The first candidate is for thread objects where,
on some architectures, it must be aligned for saving/restoring
registers.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

andrewboie commented on kernel/userspace.c:
```diff
@@ -156,6 +156,50 @@ static size_t obj_size_get(enum k_objects otype)
+
+	switch (otype) {
+	case K_OBJ_THREAD:
+		if (IS_ENABLED(CONFIG_X86_64)) {
```
I don't want arch-specific stuff like this in the core kernel.

Define some optional ARCH_THREAD_OBJECT_ALIGNMENT in arch.h as needed and use that.

dcpleung replied:
Updated to do what you suggested instead.

dcpleung committed:
x86: define dynamic thread object alignment

x86 and x86_64 require certain alignment in the k_thread struct
since the buffer to save/restore FPU/SSE registers requires
strict alignment.

Fixes #29589
Fixes #29629

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

cfriedt commented:
Looks like there is some potential conflict with #30363 which has been open for a while in preparation for dynamic pthtread stacks (#29029, which has been there for quite a long time too).

@dcpleung - I wonder if this ticket could focus more on dynamic kobject creation while #30363 could focus more on allocation.

Please let me know if you can think of a better alternative.

dcpleung commented:
> Looks like there is some potential conflict with #30363 which has been open for a while in preparation for dynamic pthtread stacks (#29029, which has been there for quite a long time too).
> 
> @dcpleung - I wonder if this ticket could focus more on dynamic kobject creation while #30363 could focus more on allocation.
> 
> Please let me know if you can think of a better alternative.

Once your PR is merged, I can rebase this on top of your changes.

nashif commented on kernel/mempool.c:
```diff
@@ -8,7 +8,7 @@
 #include <sys/math_extras.h>
 
-void *z_heap_malloc(struct k_heap *heap, size_t size)
+void *z_heap_aligned_malloc(struct k_heap *heap, size_t align, size_t size)
```
can we make this static?

dcpleung replied:
Waiting for #30363 to be merged...

cfriedt commented on kernel/include/kernel_internal.h:
```diff
@@ -46,6 +46,22 @@ extern char *z_setup_new_thread(struct k_thread *new_thread,
+ * @return A pointer to the allocated memory, or NULL if there is insufficient
+ * RAM in the pool or there is no pool to draw memory from
+ */
+void *z_thread_aligned_malloc(size_t align, size_t size);
```
It might be a bit more consistent to name it aligned_alloc rather than aligned_malloc, just for consistency.

cfriedt commented on kernel/mempool.c:
```diff
@@ -98,7 +93,7 @@ void k_thread_system_pool_assign(struct k_thread *thread)
 #endif
 
-void *z_thread_malloc(size_t size)
+void *z_thread_aligned_malloc(size_t align, size_t size)
```
It might be a bit more consistent to name it aligned_alloc rather than aligned_malloc, just for consistency.

cfriedt commented on include/sys/kobject.h:
```diff
@@ -267,12 +267,37 @@ __syscall void *k_object_alloc(enum k_objects otype);
+ */
+static inline struct z_object *z_dynamic_object_create(size_t size)
+{
+	return z_dynamic_object_aligned_create(1, size);
```
I think elsewhere that `sizeof(void*)` is used in place of an unspecified alignment because it's a reasonable safe default alignment.

cfriedt reviewed (changes requested):
Looks good - just a slight change request to have consistent naming. Are any additional tests required?

npitre commented:
On Tue, 12 Jan 2021, Christopher Friedt wrote:

> +	return z_dynamic_object_aligned_create(1, size);

> I think elsewhere that `sizeof(void*)` is used in place of an unspecified alignment because it's a reasonable safe default alignment. 

Personally, I'd prefer using 0 when no particular alignment is 
necessary. The code will already take care to return memory with the 
minimum alignment for all basic types for a given platform in that case.

dcpleung commented:
> On Tue, 12 Jan 2021, Christopher Friedt wrote: + return z_dynamic_object_aligned_create(1, size); I think elsewhere that `sizeof(void*)` is used in place of an unspecified alignment because it's a reasonable safe default alignment.
> Personally, I'd prefer using 0 when no particular alignment is necessary. The code will already take care to return memory with the minimum alignment for all basic types for a given platform in that case.

Hm... `sys_heap_aligned_alloc()` with alignment `0` simply calls `sys_heap_alloc()` which has no minimal alignment (AFAIK). So probably better to pass a value over.

dcpleung commented:
> Looks good - just a slight change request to have consistent naming. Are any additional tests required?

Updated the naming and alignment value.

npitre commented:
On Tue, 12 Jan 2021, Daniel Leung wrote:

> Hm... `sys_heap_aligned_alloc()` with alignment `0` simply calls `sys_heap_alloc()` which has no minimal alignment (AFAIK). So probably better to pass a value over.

It sure does. sys_heap_alloc() must have minimal alignment for the 
platform no matter what. And I know it does.

dcpleung commented:
> On Tue, 12 Jan 2021, Daniel Leung wrote: Hm... `sys_heap_aligned_alloc()` with alignment `0` simply calls `sys_heap_alloc()` which has no minimal alignment (AFAIK). So probably better to pass a value over.
> It sure does. sys_heap_alloc() must have minimal alignment for the platform no matter what. And I know it does.

Changed the value to `0`. There is another issue where they reported the minimal alignment was 4, but I recalled it as 1, hence the confusion.

npitre commented:
On Tue, 12 Jan 2021, Daniel Leung wrote:

> Changed the value to `0`. There is another issue where they reported the minimal alignment was 4, but I recalled it as 1, hence the confusion.

Thanks. There is a 1 in `z_thread_malloc()` that could be turned into a 0 as well for consistency.

dcpleung commented:
> On Tue, 12 Jan 2021, Daniel Leung wrote: Changed the value to `0`. There is another issue where they reported the minimal alignment was 4, but I recalled it as 1, hence the confusion.
> Thanks. There is a 1 in `z_thread_malloc()` that could be turned into a 0 as well for consistency.

Changed that too.

cfriedt reviewed (approved):
Works for me
